### PR TITLE
Adjust mechanism that moves all scripts to the end of the page

### LINKED ISSUE #26026: JsFooterPlugin sets Http Response content empty if PCRE limit reached
### Preconditions (*)
1. Magento 2.3.2-p2
2. PHP 7.1
3. pcre.backtrack_limit = 1000000 (Default)
4. pcre.recursion_limit = 1000000 (Default)
5. dev/js/move_script_to_bottom=enabled

### Steps to reproduce (*)
1. Build a huge HTML output (e.g. shopping cart with a lot of complex products)

### Expected result (*)
1. Output should be turned or error should be thrown

### Actual result (*)
1. A HTTP status code 200 without a response body is returned

The call to `preg_replace_callback()` in [`JsFooterPlugin`](https://github.com/magento/magento2/blob/9544fb243d5848a497d4ea7b88e08609376ac39e/app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php#L51) will fail silently, resulting in a blank page. The reason for this is, that `preg_replace_callback()` will return `NULL` if one of it's limits is reached. `NULL` will further be used to set the content of the HTTP response.

A first step would be to check if the result of `preg_replace_callback()` is `NULL` and throw an appropriate exception (generated from `preg_last_error()`).

### PR DESCRIPTION
**Pull request by krzksz**
### Description (*)
This PR introduces a similar change to #27211, we had this feature implemented in our custom extension and we eventually had to increase both i`pcre.backtrack_limit` and `pcre.recursion_limit` to something around `1000000` in order for the regular expressions to work for large pages.

### Fixed Issues (if relevant)

1. Possibly https://community.magento.com/t5/Magento-2-x-Technical-Issues/Magento-2-3-3-Move-JS-code-to-the-bottom-of-the-page-Error/td-p/348577
2. https://github.com/magento/magento2/issues/26026 - JsFooterPlugin sets Http Response content empty if PCRE limit reached

### Manual testing scenarios (*)

1. Change **Move JS code to the bottom of the page** to **Yes** under Stores->Configuration->Advanced->Developer.
2. Refresh the caches and open the shopfront.
3. Verify that all script tags are located in the same order right before the end of body tag.

### Questions or comments

I am unable to find any tests for this feature and I'm not experienced enough with PHP to write one from scratch so it would be great if someone contributed some tests here. Thanks!
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @krzksz,
Please review unit test failure. Other tests failures looks not related to changes from this PR

**Comment by krzksz** [comment]
Hey @ihor-sviziev, checking it now... I wonder why I was not able to find them before 🤔

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7154](https://jira.corp.magento.com/browse/ENGCOM-7154) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Comment by engcom-Echo** [comment]
Hi @krzksz,
Please review Magento Health Index test failure.

**Comment by engcom-Echo** [comment]
Failed functional tests not related to the changes in this PR

**Comment by engcom-Echo** [comment]
It should be approved. File has not inline JS.

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 1]
```diff
@@ -39,26 +39,38 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
-                    function ($matchPart) use (&$script) {
-                        $script[] = $matchPart[0];
-                        return '';
-                    },
-                    $content
-                );
-                $subject->setContent(
-                    str_replace('</body', implode("\n", $script) . "\n</body", $content)
-                );
+        $content = $subject->getContent() ?? '';
```
Comment (markshust) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: I think instead of null coalesce here we should cast this as a string:

`$content = (string) $subject->getContent();`

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 2]
```diff
@@ -39,26 +39,38 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
+
+        $bodyClose = '</body';
+
+        if (is_string($content) && strpos($content, $bodyClose) !== false && $this->scopeConfig->isSetFlag(
```
Comment (markshust) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Once the above is casted as a string, the `is_string($content)` here can be removed

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 3]
```diff
@@ -39,26 +39,38 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
+            self::XML_PATH_DEV_MOVE_JS_TO_BOTTOM,
+            ScopeInterface::SCOPE_STORE
+        )) {
+            $scripts = '';
+
+            $scriptOpen = '<script';
+            $scriptClose = '</script>';
+
+            $scriptOpenPos = strpos($content, $scriptOpen);
+            while ($scriptOpenPos !== false) {
```
Comment (markshust) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: The line breaks here are inconsistent. I think we should remove all line breaks inbetween var declarations and add one before the `while` statement.

**Review by markshust — CHANGES_REQUESTED** [review]

**Review by markshust — CHANGES_REQUESTED** [review]
I've re-read the updated code and just realized it's very hard to understand. There are a lot of string replacements, and I'm unsure what this code is supposed to be doing.

Perhaps things can be rewritten into smaller functions with more descriptive variable names? Also I'm not sure how tests would be able to be written to properly test this code -- there's too much going on in this one function.

Also notice how many references to `$scriptSomething` there are:
Screen Shot 2020-04-10 at 8 23 34 AM

**Comment by ihor-sviziev** [comment]
Hi @markshust,
I don’t agree with you there.
Implementation is really simple as possible. Yes, it’s definitely harder than using regex but it does the same in more performance optimized way from performance perspective.
Such performance improvements can’t be as simple as not optimized code. 
About test coverage - we have unit tests that checks functionality works fine.

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 4]
```diff
@@ -39,26 +39,37 @@ public function __construct(ScopeConfigInterface $scopeConfig)
      */
     public function beforeSendResponse(Http $subject)
```
Comment (markshust) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Consider some of the variable naming suggestions/methods used here. Breaking apart into two functions makes more sense to me here also as the original function is getting way too long (consider breaking it out to even more functions). The combo of these make the code much easier to reason about.

```
    /**
     * Put all javascript to footer before sending the response.
     *
     * @param Http $subject
     * @return void
     */
    public function beforeSendResponse(Http $subject)
    {
        $content = (string)$subject->getContent();
        $isEndBodyTagFound = strpos($content, '</body') !== false;
        $shouldMoveJsToBottom = $this->scopeConfig->isSetFlag(self::XML_PATH_DEV_MOVE_JS_TO_BOTTOM, ScopeInterface::SCOPE_STORE);

        if ($isEndBodyTagFound && $shouldMoveJsToBottom) {
            if ($scripts = $this->moveScriptsToBottom($content)) {
                $content = str_replace('</body', "$scripts\n</body", $content);
                $subject->setContent($content);
            }
        }
    }

    public function moveScriptsToBottom(&$content)
    {

        $scripts = '';
        $scriptOpen = '<script';
        $scriptClose = '</script>';
        $scriptOpenPos = strpos($content, $scriptOpen);

        while ($scriptOpenPos !== false) {
            $scriptClosePos = strpos($content, $scriptClose, $scriptOpenPos);
            $scriptPos = $scriptClosePos - $scriptOpenPos + strlen($scriptClose);
            $script = substr($content, $scriptOpenPos, $scriptPos);
            $isXMagentoTemplate = strpos($script, 'text/x-magento-template') !== false;

            if ($isXMagentoTemplate) {
                $scriptOpenPos = strpos($content, $scriptOpen, $scriptClosePos);
                continue;
            }

            $scripts .= "\n$script";
            $scriptContent = str_replace($script, '', $content);

            // Script cut out, continue search from its position.
            $scriptOpenPos = strpos($content, $scriptContent, $scriptOpenPos);
        }

        return $scripts;
    }
```

**Comment by markshust** [comment]
@ihor-sviziev the code is really hard to follow as written -- not sure what is going on within that function.

Please consider using strategic variable names here to help self-document the code. Perhaps an example of this will help: https://github.com/magento/magento2/pull/27270/files#r406863628

**Comment by krzksz** [comment]
Thanks for the suggestions @markshust, I'll see what I can do. There is one issue - I think unit tests are failing now because we are using string casting on an array. I'm not sure if it is possible to even get an array as an argument on the other hand.

**Comment by markshust** [comment]
@krzksz I don't want to open up a rabbit hole here, but the `$content` var is defined as a string, but casted as `mixed`:

```
class Message implements MessageInterface
{
    /**
     * @var array
     */
    protected $metadata = [];

    /**
     * @var mixed
     */
    protected $content = '';
```

I was previously under the assumption that `getContent()` return a string. Is something passing an array instead, and if so, why is it doing that? getContent should probably define a `string` return type and the `mixed` annotations should be changed to `string`.

**Comment by krzksz** [comment]
I'm not sure why it is like that and it reaches beyond my backend knowledge as a frontend dev 😄 What I am sure of is that in some cases you can get `null` instead of string.

**Comment by krzksz** [comment]
I mean, if we are certain it will be either string or `null` then I can adjust the unit tests as well.

**Comment by markshust** [comment]
I've noticed this also, and is the reason I found this ticket (had exception thrown on prod which was passing `null` around). 

These are three different types: array, string, and null (empty object). I'm not sure how null is passed around but that should never happen, if it doesn't have a value it should be retain the default value of an empty string `''`. Strings all around.

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 5]
```diff
@@ -39,26 +39,38 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
+                $content = str_replace($script, '', $content);
+                // Script cut out, continue search from its position.
+                $scriptOpenPos = strpos($content, $scriptOpen, $scriptOpenPos);
+            }
+
+            if ($scripts) {
+                $content = str_replace($bodyClose, $scripts . "\n" . $bodyClose, $content);
+                $subject->setContent($content);
             }
         }
```
Comment (mrtuvn) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php:93]: we should add return null at end of beforeSendResponse ?

**Review by mrtuvn — CHANGES_REQUESTED** [review]

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php (hunk 5 above)
Comment (lbajsarowicz) [reply app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php:93]: You don't need. No response is null response in that case

Comment (mrtuvn) [reply app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php:93]: ok seem everything good with me. Let see the final results and merge this soon

**Comment by lbajsarowicz** [comment]
I hope that Commeng will merge that soon, as I have extra idea to improve performance of this on other-than-layout outputs.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @krzksz,
Could you fix failing unit tests?

**Comment by krzksz** [comment]
Hello everyone!

- @ihor-sviziev All tests fixed.
- @markshust Code split, hope it's more readable now 🤗

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 6]
```diff
@@ -32,34 +32,57 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
+        $content = (string)$subject->getContent();
+        $bodyEndTag = '</body';
+        $isEndBodyTagFound = strpos($content, $bodyEndTag) !== false;
+        $shouldMoveJsToBottom = $this->scopeConfig->isSetFlag(
+            self::XML_PATH_DEV_MOVE_JS_TO_BOTTOM,
+            ScopeInterface::SCOPE_STORE
+        );
+
+        if ($isEndBodyTagFound && $shouldMoveJsToBottom) {
+            if ($scripts = $this->extractScriptTags($content)) {
```
Comment (ihor-sviziev) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Creating new variable inside for statement isn’t good thing as it quite often causes not clear understanding. Could you extract it to separate line, like this?
```suggestion
            $scripts = $this->extractScriptTags($content)
            if ($scripts) {
```

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 7]
```diff
@@ -32,34 +32,57 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
             }
         }
     }
+
+    /**
+     * Extracts and returns script tags found in given content.
+     *
+     * @param string $content
+     */
+    public function extractScriptTags(&$content)
```
Comment (ihor-sviziev) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Is there any reason of making this method public and not declaring return type?
```suggestion
    private function extractScriptTags(&$content): string
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php (hunk 6 above)
Comment (krzksz) [reply app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Good catch in both cases, I'll adjust the code.

**Review by markshust — DISMISSED** [review]
Looks so much better! Nice update 👍

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7154](https://jira.corp.magento.com/browse/ENGCOM-7154) has been created to process this Pull Request

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 8]
```diff
@@ -32,45 +32,52 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
+     * Moves all JavaScript tags to the end of body if this feature is enabled.
      *
      * @param Http $subject
      * @return void
      */
     public function beforeSendResponse(Http $subject)
     {
         $content = (string)$subject->getContent();
+        $bodyEndTag = '</body';
+        $isEndBodyTagFound = strpos($content, $bodyEndTag) !== false;
```
Comment (sshymko) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Note that HTML tags are case-insensitive. Replacing `strpos()` and `str_replace()` with their case-insensitive counterparts `stripos()` and `str_ireplace()` is the simplest improvement.

File: app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php [hunk 9]
```diff
@@ -32,45 +32,52 @@ public function __construct(ScopeConfigInterface $scopeConfig)
...
-            self::XML_PATH_DEV_MOVE_JS_TO_BOTTOM,
-            ScopeInterface::SCOPE_STORE
-        )) {
-            $scripts = '';
-            $scriptOpen = '<script';
-            $scriptClose = '</script>';
-            $scriptOpenPos = strpos($content, $scriptOpen);
+        if ($isEndBodyTagFound && $shouldMoveJsToBottom) {
+            if ($scripts = $this->extractScriptTags($content)) {
+                $content = str_replace($bodyEndTag, "$scripts\n$bodyEndTag", $content);
```
Comment (sshymko) [inline app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Position of the closing body tag has already been determined above. There's no need to repeat the search as part of the replacement. Script tags can be inserted at a known position.

Recommended optimization:
```php
$bodyEndTagPos = strripos($content, $bodyEndTag);
$isEndBodyTagFound = $bodyEndTagPos !== false;
// ...
$content = substr_replace($content, $scripts . "\n", $bodyEndTagPos, 0);
```

Notice the closing body tag search is performed from the end (via `strripos()`) instead of from the beginning (via `stripos()`), because it's always going to be found much much close to the end.

Comment (lbajsarowicz) [reply app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: @sshymko could we get this PR merged, and further improvements in separate one? :-)

I'm going to create one more based on Mateusz code, so you can ping me for the link

Comment (sshymko) [reply app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: Feel free to ignore my comments. I'm not part of the Magento team :)

Comment (lbajsarowicz) [reply app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: We neither. We contribute to Magento just like you without being part of the structures

Comment (lbajsarowicz) [reply app/code/Magento/Theme/Controller/Result/JsFooterPlugin.php]: @sshymko I applied your changes, the only issue I faced was the fact that previously fetched `bodyEndTagPos` changes due to `<script`s pulled out of DOM. So I'm going to slightly change the code to reflect the needs.

**Comment by lbajsarowicz** [comment]
Let's get this PR merged ASAP, and if you have further ideas for improvements - we can continue in https://github.com/magento/magento2/pull/27927

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed
All script tags are located in the same order right before the end of the body tag and JS functionality works as previously (expected).

**Comment by engcom-Echo** [comment]
Failed tests not related

**Review by slavvka — CHANGES_REQUESTED** [review]
@krzksz @engcom-Echo needs to be retested more widely since it is affects all Magento pages. Please come up with an approach how we make sure that we don't break anything.

**Comment by slavvka** [comment]
Also we need to approve the failures of MCHI tests

**Comment by lbajsarowicz** [comment]
@slavvka @engcom-Echo Also if there are no more changes reported (I expected some more), we can squash my changes to this PR before tests.

**Comment by sshymko** [comment]
@slavvka sounds like the entire test suite of functional tests needs to be executed in both modes (setting on/off), which is going to double the overhead of CI/CD.

**Comment by slavvka** [comment]
@sshymko Main concern from PO side was a potential impact at the regressions scope. Running the automated tests in both modes is a good point but could be insufficient. That's why I asked you guys to analyze changes once more and probably come up with an additional approach. It doesn't mean that this PR is rejected if you fail to come up. It is just additional work that is good to have if possible since the changes affect all Magento

**Comment by slavvka** [comment]
And the impact to CICD is not an issue since it should be done only for this PR couple of times :)

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @krzksz,
Could you bring the changes from @lbajsarowicz in https://github.com/magento/magento2/pull/27927 to your PR to make easier testing?

**Comment by lbajsarowicz** [comment]
@ihor-sviziev Added changes there.

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7154](https://jira.corp.magento.com/browse/ENGCOM-7154) has been created to process this Pull Request

**Review by mrtuvn — DISMISSED** [review]

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @krzksz @lbajsarowicz 

During testing, we faced the issue.

**_Problem:_**  After enabling the option  **Move JS code to the bottom of the page**, all scripts remain in the same places as before.

**_Precondition:_**

2.4-develop
Developer mode

**Manual testing scenario:**
1. Change **Move JS code to the bottom of the page** to **Yes** under **Stores->Configuration->Advanced->Developer.**;
2. Refresh cache;
3. Go to Storefront and open source;

**Actual Result:** :heavy_multiplication_x: 
screenshot_111
screenshot_12222
...

Could you take a look? 
Thanks!

**Comment by lbajsarowicz** [comment]
@krzksz - I'll take care of this. Sorry as it looks like something I've introduced.

**Comment by lbajsarowicz** [comment]
The issue was caused by the fact that PageCache was plugged in without `sortOrder`, so my plugin was executed AFTER caching the layout output. As a result - first time after flushing cache and without PageCache enabled - everything worked correctly, but after loading page from cache - it loaded the wrong version.

**Review by ihor-sviziev — APPROVED** [review]
Approved ✔

Functional and DB Compare tests looks like false positives.

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7154](https://jira.corp.magento.com/browse/ENGCOM-7154) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed 
All script tags are located in the same order right before the end of the body tag.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
Need to approve the failures of MCHI tests

**Comment by ihor-sviziev** [comment]
@engcom-Echo Magento Health Index seems like false positive. Are you sure we need approval for that?

**Comment by slavvka** [comment]
MCHI failure is false positive, no need to approve

**Review by slavvka — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-7154](https://jira.corp.magento.com/browse/ENGCOM-7154) has been created to process this Pull Request

**Comment by engcom-Echo** [comment]
@magento run all tests
